### PR TITLE
Use `poetry` build system, update Python version to ^3.8 to include 3.12+

### PR DESCRIPTION
**Pull request by darthtrevino**
- [x] Does this PR have a descriptive title that could go in our release notes?
- [x] Does this PR add any new dependencies?
- [x] Does this PR modify any existing APIs?
   - [x] Is the change to the API backwards compatible?
- [x] Have you built the documentation (reference and/or tutorial) and verified the generated documentation is appropriate?

#### What does this implement/fix? Briefly explain your changes.
This overall goal of this PR is to support Python 3.12+. I felt that this was not a safe thing to change without a deterministic dependency management system, so I included Poetry. The makefile scripts have been moved inside `pyproject.toml`, in addition to the content found in setup.py, setup.cfg, requirements.txt, etc.. More can be done here, but this is a first stepp.

#### Any other comments?
We'll need to address the publication step and setting a version there. There are tools and plugins we can leverage on the poetry side, but we should also be using the **Pypi Trusted Publisher** configuration to avoid having to use twine to manually upload the final artifacts.

### CODE REVIEW

File: graspologic/cluster/base.py [hunk 1]
```diff
@@ -6,11 +6,12 @@
 
 import numpy as np
 from sklearn.base import BaseEstimator, ClusterMixin
+from sklearn.cluster import AffinityPropagation
 from sklearn.metrics import adjusted_rand_score
 from sklearn.utils.validation import check_is_fitted
 
 
-class BaseCluster(ABC, BaseEstimator, ClusterMixin):
+class BaseCluster(ABC, AffinityPropagation):
```
Comment (bdpedigo) [inline graspologic/cluster/base.py:14]: could you explain this change?

Comment (darthtrevino) [reply graspologic/cluster/base.py:14]: In newer versions of sk-learn, the `affinity` argument on  BaseEstimator is no longer available, yielding testing errors. It is, hower, available on AffinityPropagation, which uses BaseEstimator and ClusterMixin as parent classes. 

BaseEstimator and ClusterMixin are removed here because they're already included via AffinityPropagation.

Comment (darthtrevino) [reply graspologic/cluster/base.py:14]: I think this can be resolved another way though, this documentation (https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering) indicates that the `metric` argument is probably what we should use when building an AgglomorativeClustering instance.

I'll update the PR with a less-intrusive change

Comment (bdpedigo) [reply graspologic/cluster/base.py]: thanks! yeah i was just looking into this and it looks like the main thing was just that kwarg change

**Comment by bdpedigo** [comment]
@darthtrevino just checking if you're all done making changes?

**Comment by bdpedigo** [comment]
also, wondering if you can describe what the new release process with this system would be. not familiar with `poetry dynamic-version`

**Comment by darthtrevino** [comment]
Yep, no more source changes in this PR.

So the poetry-dynamic-version plugin will use `git` tags to version your artifacts. When you have this plugin installed and run `poetry build` it will build the wheelfile using a version based on the most recent git tag. If you're on a git tag, it will use that tag's version. If you're on `main` and you're several commits away from a release tag, it will generate a prerelease version with a hash (e.g. 2.1.0-pre-<hash>)

So the normal release flow is like this:
* In a release PR, update the version in `pyproject.toml` to the new version (e.g. 1.6.0), this is mostly for documentation/development sanity purposes.
* When the release hits main, cut a new release in GitHub.
* The new release will create a tag (e.g. v1.2.0), the workflow will trigger and release a package with the tag version (1.2.0)

You'll probably want to ensure that the publication step will work on prerelease builds before cutting a public release.
Note that this isn't using Twine, you'll need to set up a Trusted Publisher in PyPi, which is _way_ easier: https://docs.pypi.org/trusted-publishers/

**Comment by darthtrevino** [comment]
You can test this process out locally if you've got the plugin installed - create a new tag on a branch (e.g. v4.0.0) then run `poetry build`, compare that to `poetry build` without being on a tag.

**Comment by bdpedigo** [comment]
thanks @darthtrevino !
